Input

dlindhol opened the pull request:
This still has a bit of cruft but I wanted to get something out there so people can look at it before our review tomorrow. 

The `input` stuff is similar to what it was in latis-beta (now in latis-hylatis). The big difference is the use of fs2.Streams so we can handle the IO with FP purity. I decided to bite the bullet because I really wanted the auto release of resources. Any other approach just felt like a diversion. 

The data model implementation is still evolving a bit. The main change is `SampledFunction` providing a Stream of Samples instead of an Iterator. I also changed `EvaluatableFunction` to `MemoizedFunction` since that was really the essence of the trait: all the data is loaded so you can do random access evaluation. I also decided that a `StreamFunction` (dropped the "ing" since all `SampledFunctions` are stream-able) can be evaluated within the limitations of traversing once.  

The `input` stuff required a touch of `Operations` but that remains for a later review. In particular, delegating to the adapter to handle an operation will change to be that the adapter can provide an implementation of an operation as a function that can be applied with the rest. This will allow us to preserve the original order of operations and provide opportunities to compose/compile operations.

I chose to specify the `IO` effect instead of the generic `F[_]` in the `fs2.Streams` because I ran into some hurdles. After tinkering with `MemoizedFunctions` that can be Pure, I'm going to try to go back to the generic `F`. I did a little experiment that gave me more confidence: https://github.com/latis-data/latis3/blob/input/core/src/test/scala/TestEffects.scala

dlindhol committed:
Add the latis.input package and related data model updates

lindholc commented on core/src/main/scala/latis/data/DomainOrdering.scala:
```diff
@@ -2,11 +2,17 @@ package latis.data
 
 /**
  * Define an Ordering for DomainData.
+ * DaomainData being compared must have the same number of values (arity).
```
Typo here.

lindholc commented on core/src/main/scala/latis/input/UrlStreamSource.scala:
```diff
@@ -0,0 +1,24 @@
+ */
+case class UrlStreamSource(val url: URL) extends StreamSource {
+  
+  def getStream: Stream[IO, Byte] = {
```
So, I've learned some stuff about this.

First, we do not want the global `ExecutionContext` here. Instead, we want a separate `ExecutionContext` for blocking calls. (See [here](https://gitter.im/functional-streams-for-scala/fs2?at=5bb6a5dc64cfc273f998c30d) and [here](https://gitter.im/functional-streams-for-scala/fs2?at=5bb6a666271506518dbaadc9), and the example in the [`fs2` README](https://github.com/functional-streams-for-scala/fs2), and the [`fs2` 1.0 migration guide](https://github.com/functional-streams-for-scala/fs2/blob/v1.0.0/docs/migration-guide-1.0.md#fs2io-changes).)

Second, I believe we can do something like the following:

```scala
def getStream[F[_]: Sync: ContextShift](blockingEC: ExecutionContext): Stream[F, Byte]) =
  readInputStream(Sync[F].delay(url.openStream), 4096, blockingEC)
```

When someone calls `getStream` they'll need to provide an `ExecutionContext` (_not_ the global one) and an implicit `ContextShift[F]`. I think the `ContextShift` will come from a different `ExecutionContext` than the blocking one. (Provided by `IOApp` or manually in a test or in the REPL.)

I've yet to determine if that's the best way.

Some additional background:
- [Thread Pools](https://gist.github.com/djspiewak/46b543800958cf61af6efa8e072bfd5c)
- [Best Practice: Should Not Block Threads](https://monix.io/docs/3x/best-practices/blocking.html)

lindholc replied:
I suppose we could also look at [how `readInputStream` itself is implemented](https://github.com/functional-streams-for-scala/fs2/blob/14d20f6f259d04df410dc3b1046bc843a19d73e5/io/src/main/scala/fs2/io/io.scala#L16-L26) as an example.

dlindhol replied:
The README suggests:
`val blockingEC = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(2))`
Do we want to do that for every read or reuse it "globally" (as opposed to the "global" EC)? It seems to me that if this is making a new thread pool, we should reuse it.

dlindhol commented:
I cleaned up a few things and squashed into a single commit (so don't rely on a previously checked out copy of this branch. Presumably I haven't messed up the master branch, this time.

Let me know here if you have any additional comments about the design. We can nit-pick details in the near future.
